simple_history/template_utils.py

```diff
@@ -99,23 +117,46 @@ def prepare_delta_change_value(
             display_value = value
         return display_value
 
-    def stringify_delta_change_value(self, change: ModelChange, value: Any) -> str:
+    def stringify_delta_change_values(
+        self, change: ModelChange, old: Any, new: Any
+    ) -> Tuple[SafeString, SafeString]:
         """
-        Return the displayed value for the ``old`` and ``new`` fields of ``change``,
-        after it's prepared by ``prepare_delta_change_value()``.
+        Called by ``format_delta_change()`` after ``old`` and ``new`` have been
+        prepared by ``prepare_delta_change_value()``.
 
-        :param change:
-        :param value: Either ``change.old`` or ``change.new``, as returned by
-               ``prepare_delta_change_value()``.
+        Return a tuple -- ``(old, new)`` -- where each element has been
+        escaped/sanitized and turned into strings, ready to be displayed in a template.
+        These can be HTML strings (remember to pass them through ``mark_safe()`` *after*
+        escaping).
         """
-        # If `value` is a list, stringify it using `str()` instead of `repr()`
-        # (the latter of which is the default when stringifying lists)
-        if isinstance(value, list):
-            value = f'[{", ".join(map(str, value))}]'
 
-        value = conditional_escape(value)
-        value = truncatechars_html(value, self.max_displayed_delta_change_chars)
-        return value
+        def stringify_value(value) -> Union[str, SafeString]:
+            # If `value` is a list, stringify each element using `str()` instead of
+            # `repr()` (the latter is the default when calling `list.__str__()`)
+            if isinstance(value, list):
```

Comment:
I had the same thought when writing it, but decided to only support lists to prevent unexpected behavior, since e.g. strings are instances of `Sequence`.

`value` can be of any type (returned by `prepare_delta_change_value()`), and the thought was that only M2M fields will use this code by default (since their values are lists), and that users wanting to leverage `stringify_value()`'s code could override `prepare_delta_change_value()` and return e.g. a list of objects whose `__str__()` implementation looks good enough for being displayed in the template through the `f"[{...}]"` string. *(I guess the latter point should ideally be added to the docstrings 😅)*

Reply:
With that in mind, do you have any alternative implementation suggestions..? I'm not sure I can think of any without making the code unnecessarily complicated 🤔

Reply:
Nope, that's a good enough justification to me!